src/RewardsDistributor/MorphoCompoundRewardsDistributor.ts

```diff
@@ -0,0 +1,305 @@
+      .add(event.args._balanceOnPool.mul(market.lastPoolBorrowIndex))
+      .div(constants.WeiPerEther);
+    const block = await this.blockFetcher.getBlock(event.blockNumber);
+    await this.handleBorrowSide(event.args._borrower, market, borrowBalance, block, updateRewards);
+  }
+
+  private async processRepay(event: RepaidEvent, updateRewards = true) {
+    const market = this.getOrInitMarket(event.args._poolTokenAddress);
+    const borrowBalance = event.args._balanceInP2P
+      .mul(market.lastP2PBorrowIndex)
+      .add(event.args._balanceOnPool.mul(market.lastPoolBorrowIndex))
+      .div(constants.WeiPerEther);
+    const block = await this.blockFetcher.getBlock(event.blockNumber);
+    await this.handleBorrowSide(event.args._onBehalf, market, borrowBalance, block, updateRewards);
+  }
+
+  private async handleBorrowSide(
+    userAddress: string,
+    market: MarketConfig,
+    newUserBalance: BigNumber,
+    block: providers.Block,
+    updateRewards = true
+  ) {
+    const userBalance = this.getOrInitUserBalance(userAddress, market.address);
+    if (updateRewards) {
+      market.borrowIndex = await computeBorrowIndex(market, BigNumber.from(block.timestamp), this.provider);
+      const accruedRewards = getUserAccumulatedRewards(
+        market.borrowIndex,
+        userBalance.userBorrowIndex,
+        userBalance.userBorrowBalance
+      );
+      userBalance.accruedMorpho = userBalance.accruedMorpho.add(accruedRewards);
+      userBalance.userBorrowIndex = market.borrowIndex;
+    }
+    market.lastTotalBorrow = market.lastTotalBorrow.sub(userBalance.userBorrowBalance).add(newUserBalance);
+    userBalance.userBorrowBalance = newUserBalance;
+    market.borrowUpdateBlockTimestamp = BigNumber.from(block.timestamp);
+  }
+
+  private async processP2PIndexUpdate(event: P2PIndexesUpdatedEvent) {
+    // P2PIndexesUpdated is always emitted before a transaction event (Supplied, borrowed, Withdrawn, repaid)
+    const market = this.getOrInitMarket(event.args._poolTokenAddress);
+
+    market.lastP2PBorrowIndex = event.args._p2pBorrowIndex;
+    market.lastPoolBorrowIndex = event.args._poolBorrowIndex;
+    market.lastP2PSupplyIndex = event.args._p2pSupplyIndex;
+    market.lastPoolSupplyIndex = event.args._poolSupplyIndex;
+  }
+
+  // ----------- After processing getters -----------
+
+  public async getUsersAccumulatedRewards() {
+    const finalTimestamp = this.currentEpoch?.finalTimestamp;
+    if (!finalTimestamp) {
+      throw Error("No final timestamp");
+    }
+    return Promise.all(
+      this.usersBalances.map(async ({ address, balances }) => ({
+        address,
+        accumulatedRewards: (await userBalancesToUnclaimedTokens(balances, finalTimestamp, this.provider)).toString(), // with 18 decimals
+      }))
+    );
+  }
+
+  public async totalEmitted() {
+    return this.getUsersAccumulatedRewards().then((r) =>
+      r.reduce((a, b) => a.add(b.accumulatedRewards), BigNumber.from(0))
```

Comment:
we might create a helper `sumBN` (or even extend the array prototype) that does that since we're using it quite a lot